papis/format.py

```diff
@@ -202,17 +213,19 @@ class Jinja2Formatter(Formatter):
         "{{ doc.isbn | default('ISBN-NONE', true) }}"
     """
 
+    name: ClassVar[str] = "jinja2"
+    if not exc:
+        env: Environment = Environment()  # type: ignore
+    else:
+        logger.error(
+            "The 'jinja2' formatter requires the 'jinja' library. "
+            "To use this functionality install it using e.g. "
+            "'pip install jinja2'.", exc_info=exc)
+        raise exc
```

Comment:
I'm not sure I like this "try..except" in the class body thing. It looks very hacky.

What do you think about just making this a cached function outside the class? Something like
```
@functools.lru_cache(maxsize=None)
def jinja_environment() -> Any:
	...
```
and then use that as a sort of singleton.

Reply:
Thanks for the very pretty suggestion. I think it certainly makes better use of the try-except idiom.

The only drawback is that the type annotations are slightly less precise, but I think no harm done. I wish I could say what mypy's problem with the previous arrangement was, but this has no such issue.

Reply:
to elaborate, annotating the classvar `env` with `jinja2.environment.Environment` is no longer possible because `Environment` is stuck in the singleton's scope (I think). I'm not sure if there's a good way to pass it back to the module scope.

Reply:
It should be fine to just annotate it with `Any` in this case. I'm not too bothered by that.

Reply:
I'll not admit how bothered I was. I looked into the issue and it seems python is just not aiming to support "forward reference" of types despite their best efforts. Additionally, regardless of the parser itself, mypy isn't a fan of my using either strings or NewType to attempt a stricter annotation. Any it is.